### PR TITLE
http2: refactoring to store stream headers state locally

### PR DESCRIPTION
**Pull request by birenroy**
This change tracks the HEADERS frame state of each stream internally to the StreamImpl data structure. This is intended to ease the removal of nghttp2 callbacks, as the newer Http2VisitorInterface API does not communicate the nghttp2 `hcat` state.

Commit Message: http2: refactoring to store stream headers state locally
Additional Description:
Risk Level: low; refactoring, no functional change intended
Testing: ran unit and integration tests locally
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Review by diannahu — APPROVED** [review]
Nice!

**Comment by birenroy** [comment]
@RyanTheOptimist do you know how to retry the Envoy/macOS check?

**Comment by birenroy** [comment]
Checks passed, this is ready for merge after your review, @RyanTheOptimist .

File: source/common/http/http2/codec_impl.h [hunk 1]
```diff
@@ -206,6 +206,11 @@ class ConnectionImpl : public virtual Connection,
                       public Event::DeferredDeletable,
                       public Http::MultiplexedStreamImplBase,
                       public ScopeTrackedObject {
+    enum HeadersState {
```
[the remark this answers is not in the document]
Comment (birenroy) [reply source/common/http/http2/codec_impl.h]: I made the `enum class` change. Let me know if you think a clarifying comment would still be helpful. I'm not sure what level of detail to use for commentary in this HTTP/2-specific codec glue.

**Review by RyanTheOptimist — APPROVED** [review]

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**